feat: complete the flow-integration-test contribution document (#285)
vaadin/docs: articles/guide/contributing/flow-integration-testing.asciidoc

= Integration Testing in Flow

Sometimes creating <<flow-unit-testing#,unit tests>> is not enough.
It might be important to test that the given functionality works end-to-end in an application.
This is specially important for features and bugs that depend on the browser functionality.

> Integration testing in Flow is done with a View & Integration Test combination.

Author: ```suggestion
Integration testing in Flow is being done with a View & Integration Test combination.
```

The integration tests are in link:https://github.com/vaadin/flow/tree/master/flow-tests[flow-tests] module.
Most of the integration tests for the core part are under `flow-test-core` module.

> Descriptions about integration test modules are inside link:https://github.com/vaadin/flow/blob/master/flow-tests/README.md[the read-me file in /flow-tests].

Author: ```suggestion
Descriptions about integration test modules are inside link:https://github.com/vaadin/flow/blob/master/flow-tests/README.md[the README.md file in /flow-tests].
```

> The integration tests are done with link:https://vaadin.com/testbench[TestBench], for information see <<{articles}/tools/testbench/overview#, Vaadin TestBench>>.

Reviewer: ```suggestion
The integration tests use link:https://vaadin.com/testbench[TestBench], for information see <<{articles}/tools/testbench/overview#, Vaadin TestBench>>.
```

The view `@Route` value should be the fully qualified name of the view class like `com.vaadin.flow.uitest.ui.YourTestClassNameView`.
For example:

.Example of a test view
[source, java]
----
@Route(value = "com.vaadin.flow.uitest.ui.CompositeView", layout = ViewTestLayout.class)
public class CompositeView extends AbstractDivView {
    // ...
}
----

NOTE: View class should only depend on Flow HTML components (see components in `com.vaadin.flow.component.html`), like such `NativeButton`, `Div` etc.

=== Opening the View Class in the Browser

You can open the test view in the browser by first starting up the jetty server for that module.
You can trigger the `jetty:run` Maven task for the module through your IDE,
or by running the command `mvn jetty:run -pl <test-module-name>` like `mvn jetty:run -pl flow-test-core`.

You can then open the view in the browser for example from _http://localhost:8888/view/com.vaadin.flow.uitest.ui.CompositeView_ (depending on the route used).

== Creating the Integration Test Class

The integration test class should be named the same as the `View` class that it tests.
For example`PageView` gets the test class `PageIT`.

> This will make the `open()` method find the correct test view path automatically.

Author: ```suggestion
This enables the `open()` method to find the correct test view path automatically.
```
